client/client.go

```diff
@@ -10,29 +11,79 @@ import (
 )
 
 func Start() (*ldap.Conn, error) {
-	conf, err := config.Get()
+	conf := config.Get()
+
+	ldapUrl, err := url.Parse(conf.LdapConfig.Address)
 	if err != nil {
-		slog.With("error", err).Debug("Can't get the config")
 		return nil, err
 	}
 
-	ldapUrl, err := url.Parse(conf.LdapConfig.Address)
-	if ldapUrl.Scheme == "ldap" {
-		slog.With("scheme", ldapUrl.Scheme)
-		if conf.LdapConfig.StartTLS {
-			slog.With("useStartTLS", conf.LdapConfig.StartTLS).Debug("Trying to use startTLS")
-			// We need to use startTLS
-		} else {
-			slog.With("useStartTLS", conf.LdapConfig.StartTLS).Debug("The connection will be INSECURE")
+	tlsConfig := tls.Config{InsecureSkipVerify: conf.LdapConfig.InsecureSkipVerify}
+
+	var c *ldap.Conn = nil
+
+	switch ldapUrl.Scheme {
+	case "ldap":
+		{
+			slog.With("scheme", ldapUrl.Scheme)
+			if conf.LdapConfig.StartTLS {
+				slog.With("address", conf.LdapConfig.Address).
+					Debug("Trying to connet to ldap server...")
+				c, err = ldap.DialURL(conf.LdapConfig.Address)
+				if err != nil {
+					return nil, err
+				}
+
+				slog.With("useStartTLS", conf.LdapConfig.StartTLS).
+					Debug("Trying to use startTLS")
+				err = c.StartTLS(&tlsConfig)
+				if err != nil {
+					slog.With("err", err).Debug("StartTLS has failed")
+				}
+
+				if !conf.LdapConfig.InsecureNoSSL {
+					slog.With("InsecureNoSSL", conf.LdapConfig.InsecureNoSSL).
+						Debug("Using ldap without tls. The connection will be INSECURE")
+				} else {
+					return nil, errors.New("Can't establish TLS connection and InsecureSkipVerify is false. The connection can't be secure. Aborting")
```

Comment:
Mi sa che insecure skip verify sia gestito male qua. Penso sia una opzione di tls per dire di ignorare se il certificato non puo' essere verificato con una CA (perche' e' self signed tipo il nostro e non hai la CA installata).
Quindi penso che questa parte di logica vada rifatta. Praticamente nell'else metti una connessione ldap normlae direi

Reply:
Anche perche' com'e' scritto ora non ha senso, sei nel ramo `if startTLS` e controlli lo skip verify e dici che usi ldap senza tls

Reply:
Questo l'ho copiato da Dex. La logica sarebbe: Se il server usa ldap con la porta 389 provo startTLS. Se però startTLS fallisce allo ho la possibilità di fare tutto in chiaro che però è pericoloso. Quella è l'opzione per dire "prova ad usare startTLS, se non riesci fai in chiaro". Ovviamente se è `false` appena startTLS fallisce blocca tutto.

Io l'ho implementata comunque, ma capisco se non ha senso tenerla. Dimmi te se vuoi che la rimuovo

Reply:
apposto se e' come fa dex mi fido